### PR TITLE
gh-108303: Move `test_future` into its own subdir

### PR DESCRIPTION
**Pull request by sobolevn**
Two important notes:

1. These tests are found by `test`:

```bash
» ./python.exe -m test --list-tests | grep test_future_import
test.test_future_import.test_future
test.test_future_import.test_future3
test.test_future_import.test_future4
test.test_future_import.test_future5
```

2. They do run on direct module execution:

```bash
» ./python.exe Lib/test/test_future_import/test_future.py    
........................
----------------------------------------------------------------------
Ran 24 tests in 0.087s

OK
```

* Issue: gh-108303

### CODE REVIEW

**Comment by sobolevn** [comment]
Yeap, they are executed: https://github.com/python/cpython/actions/runs/6170416321/job/16746672105?pr=109368

**Comment by vstinner** [comment]
I see that the change doesn't remove tests, but adds one test method (29 => 30 tests in total).

Before:

```
vstinner@mona$ ./python -m test test_future test_future3 test_future4 test_future5 -j0
0:00:00 load avg: 2.56 Run 4 tests in parallel using 4 worker processes
0:00:00 load avg: 2.56 [1/4] test_future5 passed
0:00:00 load avg: 2.56 [2/4] test_future3 passed
0:00:00 load avg: 2.56 [3/4] test_future4 passed
0:00:00 load avg: 2.56 [4/4] test_future passed

== Tests result: SUCCESS ==

All 4 tests OK.

Total duration: 959 ms
Total tests: run=29
Total test files: run=4/4
Result: SUCCESS
```

After:

```
vstinner@mona$ ./python -m test test_future_import -j0
0:00:00 load avg: 1.60 Run 4 tests in parallel using 4 worker processes
0:00:00 load avg: 1.60 [1/4] test_future_import.test_future3 passed
0:00:00 load avg: 1.60 [2/4] test_future_import.test_future4 passed
0:00:00 load avg: 1.60 [3/4] test_future_import.test_future5 passed
0:00:00 load avg: 1.60 [4/4] test_future_import.test_future passed

== Tests result: SUCCESS ==

All 4 tests OK.

Total duration: 947 ms
Total tests: run=30
Total test files: run=4/4
Result: SUCCESS
```

Oh, there is an additional test AFTER:

```
test.test_future_import.test_future.FutureTest.test_future4
```

It's this test:

```py
    def test_future4(self):
        with import_helper.CleanImport('test.test_future_import.test_future4'):
            from test.test_future_import import test_future4
```

I don't understand the design of these tests. ``test_future.test_future4()`` just imports ``test_future4()``, but it doesn't execute it? Usually, we don't use "test_xxx.py"  files for that, but "xxx.py". It's just surprising.

**Comment by vstinner** [comment]
I'm not sure about ``test_future_import`` name. I would prefer to stick to just ``test_future``. If you want to keep "import", I would prefer ``test_import_future`` :-)

``Lib/test/test___future__.py`` should be included in this directory as well, no?

**Comment by sobolevn** [comment]
`test_future` is problematic. I faced several test failures because of `Lib/test/test_concurrent_futures/test_future.py` and `Lib/test/test_future/test_future.py`. This might be a bug in `libregrtest`. I will report it, but I will go with `test_import_future` for now.

> test.test_future_import.test_future.FutureTest.test_future4

Yes, it was missing, I've decided to add it. All other tests were imported like this, except this one. I also a bit confused about the design of these tests, but I don't see any harm in them as well.

> test___future__.py

Indeed, thanks for the suggestion!

**Comment by vstinner** [comment]
> I will go with test_import_future for now.

The documentation calls it "Future **statement** definitions": https://docs.python.org/dev/library/__future__.html

It's more a statement than a module, *even if* a real module exists.

Maybe use ``test_future_statement`` or  ``test_future_stmt`` name.

**Comment by sobolevn** [comment]
I've created a new issue to fix `test_future/` folder name bug: https://github.com/python/cpython/issues/109402

**Comment by sobolevn** [comment]
👍 `test_future_stmt`

**Comment by vstinner** [comment]
Honestly, the test filenames are not very helpful :-( While we are moving code and renaming files anyway, would it be possible to come up with better names?

```
$ cd Lib/test/test_future_stmt/
$ ls test_*.py -1
test_future3.py
test_future4.py
test_future5.py
test___future__.py
test_future.py
```

**Comment by sobolevn** [comment]
How about these names?

```
» ./python.exe -m test test_future_stmt
0:00:00 load avg: 1.47 Run 5 tests sequentially
0:00:00 load avg: 1.47 [1/5] test_future_stmt.test_future
0:00:00 load avg: 1.47 [2/5] test_future_stmt.test_future_flags  (test___future__)
0:00:00 load avg: 1.47 [3/5] test_future_stmt.test_future_multiple_features (test_future5)
0:00:00 load avg: 1.47 [4/5] test_future_stmt.test_future_multiple_imports (test_future4)
0:00:00 load avg: 1.47 [5/5] test_future_stmt.test_future_single_import (test_future3)
```

**Comment by vstinner** [comment]
> How about these names?

You look different and more useful, thanks! Can you update your PR please?

File: Lib/test/test_future_stmt/test_future.py [hunk 1]
```diff
@@ -25,57 +25,71 @@ def check_syntax_error(self, err, basename, lineno, offset=1):
         self.assertEqual(err.offset, offset)
 
     def test_future1(self):
```
Comment (vstinner) [inline Lib/test/test_future_stmt/test_future.py:27]: test_future1 has an interesting comment that you can copy there:

```
# Import the name nested_scopes twice to trigger SF bug #407394 (regression).
```

I suggest to rename the file and test to: ``test_import_nested_scope_twice``. For the filename, just drop ``test_`` prefix: ``import_nested_scope_twice.py``?

File: Lib/test/test_future_stmt/test_future.py [hunk 2]
```diff
@@ -25,57 +25,71 @@ def check_syntax_error(self, err, basename, lineno, offset=1):
...
-        with import_helper.CleanImport('future_test1'):
-            from test import future_test1
+        with import_helper.CleanImport('test.test_future_stmt.future_test1'):
+            from test.test_future_stmt import future_test1
             self.assertEqual(future_test1.result, 6)
 
     def test_future2(self):
```
Comment (vstinner) [inline Lib/test/test_future_stmt/test_future.py:32]: I suggest to rename the file and test to: ``test_nested_scope``.

File: Lib/test/test_future_stmt/test_future.py [hunk 3]
```diff
@@ -25,57 +25,71 @@ def check_syntax_error(self, err, basename, lineno, offset=1):
...
+        ):
+            from test.test_future_stmt import test_future_multiple_imports
+
+    def test_future_multiple_features(self):
+        with import_helper.CleanImport(
+            "test.test_future_stmt.test_future_multiple_features",
+        ):
+            from test.test_future_stmt import test_future_multiple_features
 
     def test_badfuture3(self):
```
Comment (vstinner) [inline Lib/test/test_future_stmt/test_future.py:55]: This one tests that a typo is catched properly, that an invalid name raises SyntaxError: ``from __future__ import rested_snopes``.

Maybe: ``test_unknown_feature``. The doc calls ``__future__`` names as "features": https://docs.python.org/dev/library/__future__.html

File: Lib/test/test_future_stmt/test_future.py [hunk 4]
```diff
@@ -25,57 +25,71 @@ def check_syntax_error(self, err, basename, lineno, offset=1):
...
         with self.assertRaises(SyntaxError) as cm:
-            from test import badsyntax_future3
+            from test.test_future_stmt import badsyntax_future3
         self.check_syntax_error(cm.exception, "badsyntax_future3", 3)
 
     def test_badfuture4(self):
```
Comment (vstinner) [inline Lib/test/test_future_stmt/test_future.py:60]:
```
$ python badsyntax_future4.py 
SyntaxError: from __future__ imports must occur at the beginning of the file
```

File: Lib/test/test_future_stmt/test_future.py [hunk 5]
```diff
@@ -25,57 +25,71 @@ def check_syntax_error(self, err, basename, lineno, offset=1):
...
         with self.assertRaises(SyntaxError) as cm:
-            from test import badsyntax_future4
+            from test.test_future_stmt import badsyntax_future4
         self.check_syntax_error(cm.exception, "badsyntax_future4", 3)
 
     def test_badfuture5(self):
```
Comment (vstinner) [inline Lib/test/test_future_stmt/test_future.py:65]:
```
SyntaxError: from __future__ imports must occur at the beginning of the file
```

import fails if done after another import.

File: Lib/test/test_future_stmt/test_future.py [hunk 6]
```diff
@@ -25,57 +25,71 @@ def check_syntax_error(self, err, basename, lineno, offset=1):
...
         with self.assertRaises(SyntaxError) as cm:
-            from test import badsyntax_future5
+            from test.test_future_stmt import badsyntax_future5
         self.check_syntax_error(cm.exception, "badsyntax_future5", 4)
 
     def test_badfuture6(self):
```
Comment (vstinner) [inline Lib/test/test_future_stmt/test_future.py:70]:
```
SyntaxError: from __future__ imports must occur at the beginning of the file
```

Import after a statement.

File: Lib/test/test_future_stmt/test_future.py [hunk 7]
```diff
@@ -25,57 +25,71 @@ def check_syntax_error(self, err, basename, lineno, offset=1):
...
         with self.assertRaises(SyntaxError) as cm:
-            from test import badsyntax_future6
+            from test.test_future_stmt import badsyntax_future6
         self.check_syntax_error(cm.exception, "badsyntax_future6", 3)
 
     def test_badfuture7(self):
```
Comment (vstinner) [inline Lib/test/test_future_stmt/test_future.py:75]: ... Maybe these tests would be more explicit if they would not use a file, but pass the code as a string (maybe use textwrap.dedent()).

**Review by vstinner — APPROVED** [review]
LGTM.

The tests could be *enhanced* to have a better name and description (documentation/comment/docstring) than just... a counter: test1, test2, test3, ...

Maybe using inline code rather than loading file would help to make tests more explicit.

I don't want to hold this PR further. Enhancing the test can be done later.

**Comment by vstinner** [comment]
Merged, thanks for this nice change!

Feel free to ignore my remarks on finding better name/doc for tests ;-)

**Comment by sobolevn** [comment]
I will open a PR with better names later today :)
Thanks a lot, Victor!

**Pull request merged**